src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java

```diff
@@ -637,15 +645,159 @@ private int putValuesOfAtt(String name, Datatype dataType, int index) throws Til
+          long[] buffConverted = Arrays.stream(buff).map(i -> i * 60 * 1000000).toArray();
+          if (resultVectors.length > 0) {
+            resultVectors[index].reset();
+            resultVectors[index].putLongs(0, bufferLength, buffConverted, 0);
+          }
+          break;
+        }
+      case TILEDB_DATETIME_HR:
+        {
+          long[] buff = (long[]) query.getBuffer(name);
+          bufferLength = buff.length;
+          numValues = bufferLength;
+          long[] buffConverted = Arrays.stream(buff).map(i -> i * 60 * 60 * 1000000).toArray();
+          if (resultVectors.length > 0) {
+            resultVectors[index].reset();
+            resultVectors[index].putLongs(0, bufferLength, buffConverted, 0);
+          }
+          break;
+        }
       case TILEDB_DATETIME_DAY:
         {
           long[] buff = (long[]) query.getBuffer(name);
           bufferLength = buff.length;
-          int[] buffConverted = Arrays.stream(buff).mapToInt(i -> ((Long) i).intValue()).toArray();
           numValues = bufferLength;
+          for (int i = 0; i < buff.length; i++) {
+            ms = zeroDateTime.toInstant().atOffset(ZoneOffset.UTC).plusDays(buff[i]);
+            buff[i] = ChronoUnit.MICROS.between(zeroDateTime, ms);
+          }
+          if (resultVectors.length > 0) {
+            resultVectors[index].reset();
+            resultVectors[index].putLongs(0, bufferLength, buff, 0);
+          }
+          break;
+        }
+      case TILEDB_DATETIME_WEEK:
+        {
+          long[] buff = (long[]) query.getBuffer(name);
+          bufferLength = buff.length;
+          numValues = bufferLength;
+          for (int i = 0; i < buff.length; i++) {
+            ms = zeroDateTime.toInstant().atOffset(ZoneOffset.UTC).plusWeeks(buff[i]);
+            buff[i] = ChronoUnit.MICROS.between(zeroDateTime, ms);
+          }
+          if (resultVectors.length > 0) {
+            resultVectors[index].reset();
+            resultVectors[index].putLongs(0, bufferLength, buff, 0);
+          }
+          break;
+        }
+      case TILEDB_DATETIME_MONTH:
+        {
+          long[] buff = (long[]) query.getBuffer(name);
+          bufferLength = buff.length;
+          numValues = bufferLength;
+          for (int i = 0; i < buff.length; i++) {
+            ms = zeroDateTime.toInstant().atOffset(ZoneOffset.UTC).plusMonths(buff[i]);
+            buff[i] = ChronoUnit.MICROS.between(zeroDateTime, ms);
+          }
+          if (resultVectors.length > 0) {
+            resultVectors[index].reset();
+            resultVectors[index].putLongs(0, bufferLength, buff, 0);
+          }
+          break;
+        }
+      case TILEDB_DATETIME_YEAR:
+        {
+          long[] buff = (long[]) query.getBuffer(name);
+          bufferLength = buff.length;
+          numValues = bufferLength;
+          for (int i = 0; i < buff.length; i++) {
+            ms = zeroDateTime.toInstant().atOffset(ZoneOffset.UTC).plusYears(buff[i]);
```

Comment:
Could we improve this if we move `toInstant().atOffset(ZoneOffset.UTC)` outside of the loop? That is, `private static final OffsetDateTime zeroDateTime =
      new Timestamp(0).toInstant().atOffset(ZoneOffset.UTC)`.